File: src/repositories/customer.repository.ts
import { Collection, ObjectId, ClientSession } from "mongodb";
import { getDatabaseForCompany } from "@/config/database";
import { CustomerDb } from "@/types/customer/customer";
import { logger } from "@/utils/logger";

export class CustomerRepository {
  private getCollection(companyId: string): Collection<CustomerDb> {
    const db = getDatabaseForCompany(companyId);
    return db.collection<CustomerDb>("customers");
  }

  async create(companyId: string, customer: CustomerDb): Promise<CustomerDb> {
    try {
      const collection = this.getCollection(companyId);
      await collection.insertOne(customer as any);
      logger.info("Customer created", { customerId: customer._id, companyId });
      return customer;
    } catch (error) {
      logger.error("Failed to create customer", error);
      throw error;
    }
  }

  async findById(companyId: string, id: string): Promise<CustomerDb | null> {
    try {
      const collection = this.getCollection(companyId);
      const customer = await collection.findOne({ _id: new ObjectId(id) } as any);
      return customer;
    } catch (error) {
      logger.error("Failed to find customer by ID", error);
      throw error;
    }
  }

  async getAll(companyId: string): Promise<CustomerDb[]> {
    try {
      const collection = this.getCollection(companyId);
      const customers = await collection
        .find({ status: "active" }) // Only active customers by default for dropdowns etc.
        .sort({ name: 1, surname: 1 })
        .toArray();

      return customers;
    } catch (error) {
      logger.error("Failed to fetch all customers", error);
      throw error;
    }
  }

  async findAll(
    companyId: string,
    options: {
      page: number;
      limit: number;
      status?: string;
      search?: string;
    }
  ): Promise<{ customers: CustomerDb[]; total: number }> {
    try {
      const collection = this.getCollection(companyId);
      const { page, limit, status, search } = options;
      const skip = (page - 1) * limit;

      // Build query
      const query: any = {};

      if (status) {
        query.status = status;
      }

      if (search) {
        query.$or = [
          { name: { $regex: search, $options: "i" } },
          { surname: { $regex: search, $options: "i" } },
        ];
      }

      // Execute queries in parallel
      const [customers, total] = await Promise.all([
        collection
          .find(query)
          .sort({ createdAt: -1 })
          .skip(skip)
          .limit(limit)
          .toArray(),
        collection.countDocuments(query),
      ]);

      logger.debug("Customers fetched", { companyId, count: customers.length, total });

      return { customers, total };
    } catch (error) {
      logger.error("Failed to fetch customers", error);
      throw error;
    }
  }

  async update(
    companyId: string,
    id: string,
    updates: Partial<CustomerDb>
  ): Promise<CustomerDb | null> {
    try {
      const collection = this.getCollection(companyId);
      const result = await collection.findOneAndUpdate(
        { _id: new ObjectId(id) } as any,
        { $set: { ...updates, updatedAt: new Date() } },
        { returnDocument: "after" }
      );

      if (result) {
        logger.info("Customer updated", { customerId: id, companyId });
      }

      return result;
    } catch (error) {
      logger.error("Failed to update customer", error);
      throw error;
    }
  }

  async delete(companyId: string, id: string, session?: ClientSession): Promise<boolean> {
    try {
      const collection = this.getCollection(companyId);
      const result = await collection.deleteOne({ _id: new ObjectId(id) } as any, { session });
      const deleted = result.deletedCount > 0;

      if (deleted) {
        logger.info("Customer deleted", { customerId: id, companyId });
      }

      return deleted;
    } catch (error) {
      logger.error("Failed to delete customer", error);
      throw error;
    }
  }

  async exists(companyId: string, id: string): Promise<boolean> {
    try {
      const collection = this.getCollection(companyId);
      const count = await collection.countDocuments({ _id: new ObjectId(id) } as any);
      return count > 0;
    } catch (error) {
      logger.error("Failed to check customer existence", error);
      throw error;
    }
  }
}

File: fisoft/src/main/kotlin/repository/CustomerRepository.kt
package com.ilyasdemirkiran.repository

import com.ilyasdemirkiran.config.Database
import com.ilyasdemirkiran.types.customers.Customer
import com.mongodb.client.model.Filters
import com.mongodb.client.model.FindOneAndUpdateOptions
import com.mongodb.client.model.ReturnDocument
import com.mongodb.client.model.Updates
import com.mongodb.kotlin.client.coroutine.MongoCollection
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.toList
import org.bson.Document
import org.bson.types.ObjectId

class CustomerRepository {
    private fun getCollection(companyId: String): MongoCollection<Customer> {
        return Database.getDatabaseForCompany(companyId).getCollection<Customer>("customers")
    }

    suspend fun create(companyId: String, customer: Customer): Customer {
        val id = customer._id ?: ObjectId().toHexString()
        val created = customer.copy(_id = id)
        getCollection(companyId).insertOne(created)
        return created
    }

    suspend fun findById(companyId: String, id: String): Customer? {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        return getCollection(companyId).find(filter).firstOrNull()
    }

    suspend fun getAll(companyId: String, status: String? = null): List<Customer> {
        val query = if (status != null) Filters.eq("status", status) else Document()
        return getCollection(companyId).find(query).toList()
    }

    suspend fun findAll(companyId: String, page: Int, limit: Int, status: String? = null, search: String? = null): Pair<List<Customer>, Long> {
        val filters = mutableListOf<org.bson.conversions.Bson>()
        if (status != null) filters.add(Filters.eq("status", status))
        if (!search.isNullOrEmpty()) {
            filters.add(
                Filters.or(
                    Filters.regex("name", search, "i"),
                    Filters.regex("surname", search, "i")
                )
            )
        }

        val filter = if (filters.isNotEmpty()) Filters.and(filters) else Document()
        val skip = (page - 1) * limit
        val collection = getCollection(companyId)

        val total = collection.countDocuments(filter)
        val customers = collection.find(filter)
            .sort(Document("createdAt", -1))
            .skip(skip)
            .limit(limit)
            .toList()

        return Pair(customers, total)
    }

    suspend fun update(companyId: String, id: String, name: String? = null, surname: String? = null, phone: String? = null, city: String? = null, district: String? = null, address: String? = null, status: String? = null): Customer? {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        val updates = mutableListOf<org.bson.conversions.Bson>()
        if (name != null) updates.add(Updates.set("name", name))
        if (surname != null) updates.add(Updates.set("surname", surname))
        if (phone != null) updates.add(Updates.set("phone", phone))
        if (city != null) updates.add(Updates.set("city", city))
        if (district != null) updates.add(Updates.set("district", district))
        if (address != null) updates.add(Updates.set("address", address))
        if (status != null) updates.add(Updates.set("status", status))

        if (updates.isEmpty()) return findById(companyId, id)

        return getCollection(companyId).findOneAndUpdate(
            filter,
            Updates.combine(updates),
            FindOneAndUpdateOptions().returnDocument(ReturnDocument.AFTER)
        )
    }

    suspend fun delete(companyId: String, id: String): Boolean {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        val res = getCollection(companyId).deleteOne(filter)
        return res.deletedCount > 0
    }

    suspend fun exists(companyId: String, id: String): Boolean {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        val count = getCollection(companyId).countDocuments(filter)
        return count > 0
    }
}